### PR TITLE
[SQL] Aggregations refactor

### PR DESCRIPTION
**Pull request by akedin**
Detangling and documenting some code related to aggregations.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by akedin** [comment]
R: @apilloud @kennknowles @amaliujia 

No new stuff, just splitting some classes out

**Comment by akedin** [comment]
run java postcommit

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamBuiltinAggregations.java [hunk 1]
```diff
@@ -17,96 +17,120 @@
...
-      case INTEGER:
+      case BYTE:
+      case INT16:
+      case FLOAT:
+      case DATETIME:
+      case DECIMAL:
+        return new CustMin();
+      case INT32:
         return Min.ofIntegers();
-      case SMALLINT:
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamBuiltinAggregations.java]: Is this fixing a bug that Schema has no longer SMALLINT type (and other integer types)?

Can we create a JIRA to improve the test coverage on aggregation such that we could discover the breaking changes.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamAggregationRel.java [hunk 2]
```diff
@@ -113,10 +117,59 @@ public RelWriter explainTerms(RelWriter pw) {
...
+            .stream()
+            .map(aggCall -> AggregationCombineFnAdapter.of(aggCall, inputSchema))
+            .collect(toList());
+
+    return new Transform(
+        windowFn, windowFieldIndex, inputSchema, getGroupSet(), aggregationAdapters, outputSchema);
   }
 
-  private class Transform extends PTransform<PCollectionList<Row>, PCollection<Row>> {
+  private static class Transform extends PTransform<PCollectionList<Row>, PCollection<Row>> {
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamAggregationRel.java]: I think making this class static is not necessary even if you want to pass a bunch of parameters to Transform. Seems like the main benefit of static inner class is not requiring creating an outer class's instance, which does not apply here.

https://stackoverflow.com/a/22098846/10055573

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamBuiltinAggregations.java (hunk 1 above)
Comment (akedin) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamBuiltinAggregations.java]: It supports everything, I just changed it from Calcite types to Beam Schema types

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamAggregationRel.java (hunk 2 above)
Comment (akedin) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamAggregationRel.java]: It is not necessary but it helps, e.g. if you want to embed lambdas then it's easy to have something non-serializable or something that you don't actually want to serialize in the closure. Errors that we throw in these cases are non-informational, making it static and keeping only things that are needed by the class in the class itself makes it easier to trace everything

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamBuiltinAggregations.java (hunk 1 above)
Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamBuiltinAggregations.java]: I see. Didn't aware that there is a conversion of types.

Comment (akedin) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamBuiltinAggregations.java]: There's no type conversion here, here we choose a specific implementation of the combiner based on the field type, e.g. `Sum.ofIntegers()` vs `Sum.ofLongs()` and similar. Calcite types are not serializable and as soon as you need to pass them around, things get complicated.

Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamBuiltinAggregations.java]: Gotcha.

**Comment by amaliujia** [comment]
LGTM

**Pull request merged**